pkg/controller/persistentvolume/state.go

```diff
@@ -0,0 +1,307 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package persistentvolume
+
+import (
+	"fmt"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/errors"
+)
+
+/*
+PersistentVolumes and PersistentVolumeClaims have closely related functionality and there is a lot of interplay
+between these two objects, from binding to provisioning to recycling.  In order to help manage the complexity
+inherent in the PersistentVolumeController, the State pattern (https://en.wikipedia.org/wiki/State_pattern) has
+been implemented for both PV and PVC.  This allows for easier testing with small focused methods that are
+ easier to debug.
+
+Each Phase a PV/PVC has its own State implementation.
+
+*/
+
+type persistentVolumeStateContext struct {
+	state         persistentVolumeState
+	volume        *api.PersistentVolume
+	claim         *api.PersistentVolumeClaim
+	claimNotFound bool
+	client        binderClient
+	volumeIndex   *persistentVolumeOrderedIndex
+}
+
+type persistentVolumeClaimStateContext struct {
+	state persistentVolumeClaimState
+	claim *api.PersistentVolumeClaim
+}
+
+type persistentVolumeState interface {
+	saveStatus(ctx *persistentVolumeStateContext) error
```

Comment:
bind and provision work on the volume spec itself while saving status is a separate API call.  saving status can advance to the next phase or perhaps we want a more explicit `nextPhase` func, but it wouldn't always advance (just as saving status wouldn't always have to save).